### PR TITLE
Alukavenko

### PR DESCRIPTION
**Pull request by promahos**
Rozetka tests: two separate tests - one for google search, one for Rozetka itself

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    @FindAll({@FindBy(xpath = "//cite/../../h3")})
+    private List<WebElement> searchResults;
+
+    public List<String> getSearchResultsText() {
+        return searchResults.stream()
+                .filter(WebElement::isDisplayed)
+                .map(WebElement::getText)
+                .collect(Collectors.toList());
+    }
+
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java:21]: delete new line

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class GoogleSearchResultsPage extends BasePage {
+
+    @FindAll({@FindBy(xpath = "//cite/../../h3")})
+    private List<WebElement> searchResults;
+
+    public List<String> getSearchResultsText() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java:15]: getSearchResultTitles

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleMainPage.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    @FindBy(name = "q")
+    private WebElement searchField;
+
+    @FindBy(xpath = "//input[@name = 'btnK']")
+    private WebElement searchButton;
+
+    public GoogleMainPage() {
+        driver.get("https://www.google.com/");
+        PageFactory.initElements(driver, this);
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleMainPage.java:17]: same

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaBTPage.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+public class RozetkaBTPage extends BasePage {
+
+    @FindAll({@FindBy(xpath = "//div[@class='tile-cats']")})
+    private List<WebElement> subCategories;
+
+    public int getSubCategoriesQuantity() {
+        return subCategories.size();
+    }
+
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaBTPage.java:17]: same

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    @FindBy(xpath = "//a[@class = 'menu-categories__link' and @href='https://bt.rozetka.com.ua/']")
+    private WebElement categoryBT;
+
+    public RozetkaMainPage() {
+        driver.get("https://rozetka.com.ua/");
+        PageFactory.initElements(driver, this);
+    }
+
+    public RozetkaBTPage goToCategoryBT() {
+
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java:18]: delete empty line

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import org.openqa.selenium.support.PageFactory;
+
+public class RozetkaMainPage extends BasePage {
+
+    @FindBy(xpath = "//a[@class = 'menu-categories__link' and @href='https://bt.rozetka.com.ua/']")
+    private WebElement categoryBT;
+
+    public RozetkaMainPage() {
+        driver.get("https://rozetka.com.ua/");
+        PageFactory.initElements(driver, this);
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java:14]: same

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public RozetkaBTPage goToCategoryBT() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java:17]: goToHomeAppliancesCategory

File: src/test/java/com/softserveinc/ita/alukavenko/BaseTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.softserveinc.ita.pageobjects.alukavenko.BasePage;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class BaseTest {
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/alukavenko/BaseTest.java:11]: TestRunner

move to utils package

File: src/test/java/com/softserveinc/ita/alukavenko/RozetkaSimpleTest.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        driver.manage().window().maximize();
+    }
+
+    @Test
+    public void verifyGoogleSearch() {
+
+        driver.get("https://www.google.com/");
+        driver.findElement(name("q")).sendKeys("Rozetka");
+        driver.findElement(xpath("//input[@name = 'btnK']")).click();
+
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/alukavenko/RozetkaSimpleTest.java:35]: pay attention to new lines and double new lines

File: src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+public class RozetkaSubCategoryTest extends BaseTest {
+
+    private static final int EXPECTED_SUB_CATEGORY_QUANTITY = 10;
+
+    @Test
+    public void verifyBTSubCategoryQuantity() {
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java:13]: rename

Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java:13]: rename

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java (hunk 1 above)
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        return searchResults.stream()
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java (hunk 2 above)
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+                .filter(WebElement::isDisplayed)
+                .map(WebElement::getText)
+                .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleSearchResultsPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleMainPage.java (hunk 3 above)
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/GoogleMainPage.java]: changed to static import

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaBTPage.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package com.softserveinc.ita.pageobjects.alukavenko;
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindAll;
+import org.openqa.selenium.support.FindBy;
+
+import java.util.List;
+
+public class RozetkaBTPage extends BasePage {
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaBTPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java]: done

Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java]: done

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java (hunk 6 above)
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java]: changed to static import

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java]: renamed

Comment (promahos) [reply src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java]: renamed

File: src/test/java/com/softserveinc/ita/alukavenko/BaseTest.java (hunk 8 above)
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/BaseTest.java]: Done, but isn't it strange that tests classed extend TestRunner class?
Sounds a little bit confusing to me.

Also I wasn't sure where to create utils package - either in main branch or in test. 
Utils sounds like something that is common for the project, not tests only. But in that case we need at least to change the dependecies, in order to testng was availible not only in the test scope. So for now I decided to create utils package in the test branch.

[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/BaseTest.java]: done

File: src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+import static org.testng.Assert.assertTrue;
+
+public class GoogleSearchTest extends BaseTest {
+    private static final String SEARCH_STRING = "rozetka";
+    private static final String EXPECTED_STRING = "rozetka";
+
+    @Test
+    public void verifySearchResults() {
+
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java]: done

File: src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        List<String> searchTexts = new GoogleMainPage()
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java:18]: thanks for the advice, but if it's not obligatory - I'd rather not change it now
I have some difficulity getting used to Java's strong typing, so I'd prefer to continue the "strict way". After I feel comfortable with it, I'll start using "simplifiers" like var.

[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java:18]: duplicate

File: src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+                .search(SEARCH_STRING)
+                .getSearchResultsText();
+
+        assertTrue(searchTexts.stream().allMatch(t-> t.toLowerCase().contains(EXPECTED_STRING)),
+                "Search results should contain: " + EXPECTED_STRING
+                        + ". Actual results: \n" + String.join("\n", searchTexts));
+
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java:25]: done

[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java:25]: done

File: src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    @Test
+    public void verifySearchResults() {
+
+        List<String> searchTexts = new GoogleMainPage()
+                .search(SEARCH_STRING)
+                .getSearchResultsText();
+
+        assertTrue(searchTexts.stream().allMatch(t-> t.toLowerCase().contains(EXPECTED_STRING)),
+                "Search results should contain: " + EXPECTED_STRING
+                        + ". Actual results: \n" + String.join("\n", searchTexts));
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java]: done

File: src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    private static final String EXPECTED_STRING = "rozetka";
+
+    @Test
+    public void verifySearchResults() {
+
+        List<String> searchTexts = new GoogleMainPage()
+                .search(SEARCH_STRING)
+                .getSearchResultsText();
+
+        assertTrue(searchTexts.stream().allMatch(t-> t.toLowerCase().contains(EXPECTED_STRING)),
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java]: done

[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/GoogleSearchTest.java]: done

File: src/test/java/com/softserveinc/ita/alukavenko/RozetkaSimpleTest.java (hunk 9 above)
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/RozetkaSimpleTest.java]: checked

File: src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
```
[the remark this answers is not in the document]
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java]: done

File: src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java (hunk 10 above)
Comment (promahos) [reply src/test/java/com/softserveinc/ita/alukavenko/RozetkaSubCategoryTest.java]: renamed

File: pom.xml [hunk 20]
```diff
@@ -27,6 +27,12 @@
             <artifactId>selenium-java</artifactId>
             <version>${selenium.java.version}</version>
         </dependency>
+        <!-- https://mvnrepository.com/artifact/io.github.bonigarcia/webdrivermanager -->
```
Comment (Nikodych) [inline pom.xml:30]: if you are adding such a line, do it for all libs dependencies

File: src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java [hunk 21]
```diff
@@ -2,22 +2,21 @@
...
     }
 
-    public RozetkaBTPage goToCategoryBT() {
-
-        categoryBT.click();
-
-        return new RozetkaBTPage();
+    public RozetkaHomeAppliancesPage goToHomeAppliancesCategory() {
+        homeAppliancesCategory.click();
+        return new RozetkaHomeAppliancesPage();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/alukavenko/RozetkaMainPage.java:20]: if there is a code before the return, add new line between

**Review by Nikodych — CHANGES_REQUESTED** [review]

**Pull request merged**